src/geometry/utility/geometry.cc

```diff
@@ -0,0 +1,308 @@
+}
+
+maliput::math::Vector3 MakeCenterpoint(const maliput::math::Vector3& p1, const maliput::math::Vector3& p2) {
+  return (p1 + p2) / 2.;
+}
+
+// Helper class for evaluating the boundaries of a lane.
+class BoundChecker {
+ public:
+  BoundChecker(const LineString3d& left, const LineString3d& right)
+      : left_(left), right_(right), entry_{right.first(), left.first()}, exit_{left.last(), right.last()} {
+    left_segments_ = MakeSegments(left_);
+    right_segments_ = MakeSegments(right_);
+  }
+
+  bool Intersects2d(const Segment3d& seg) const {
+    const Segment2d seg2d{{seg.first.x(), seg.first.y()}, {seg.second.x(), seg.second.y()}};
+    const bool result =
+        IntersectsLeft2d(seg2d) || IntersectsRight2d(seg2d) || CrossesEntry2d(seg2d) || CrossesExit2d(seg2d);
+    return result;
+  }
+
+  bool IntersectsLeft2d(const Segment2d& seg) const {
+    for (const auto left_segment : left_segments_) {
+      const Segment2d left_seg{{left_segment.first.x(), left_segment.first.y()},
+                               {left_segment.second.x(), left_segment.second.y()}};
+      if (SegmentsIntersect2d(seg, left_seg)) {
+        if (!(seg.first == left_seg.first)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  bool IntersectsRight2d(const Segment2d& seg) const {
+    for (const auto right_segment : right_segments_) {
+      const Segment2d right_seg{{right_segment.first.x(), right_segment.first.y()},
+                                {right_segment.second.x(), right_segment.second.y()}};
+      if (SegmentsIntersect2d(right_seg, seg)) {
+        if (!(seg.first == right_seg.first)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  bool CrossesEntry2d(const Segment2d& seg) const {
+    const Segment2d entry_2d{{entry_.first.x(), entry_.first.y()}, {entry_.second.x(), entry_.second.y()}};
+    if (SegmentsIntersect2d(seg, entry_2d)) {
+      return PointIsLeftOf({entry_2d.first.x(), entry_2d.first.y(), 0.}, {entry_2d.second.x(), entry_2d.second.y(), 0.},
+                           {seg.second.x(), seg.second.y(), 0.});
+    }
+    return false;
+  }
+
+  bool CrossesExit2d(const Segment2d& seg) const {
+    const Segment2d exit_2d{{exit_.first.x(), exit_.first.y()}, {exit_.second.x(), exit_.second.y()}};
+    if (SegmentsIntersect2d(seg, exit_2d)) {
+      PointIsLeftOf({exit_2d.first.x(), exit_2d.first.y(), 0.}, {exit_2d.second.x(), exit_2d.second.y(), 0.},
+                    {seg.second.x(), seg.second.y(), 0.});
+    }
+    return false;
+  }
+
+  template <typename Func>
+  void ForEachPointLeftUntil(LineString3d::const_iterator iter, Func&& f) const {
+    return ForEachPointUntilImpl(iter, left_, std::forward<Func>(f));
+  }
+  template <typename Func>
+  void ForEachPointRightUntil(LineString3d::const_iterator iter, Func&& f) const {
+    return ForEachPointUntilImpl(iter, right_, std::forward<Func>(f));
+  }
```

Comment:
I highly suggest you don't provide this extra layer and instead make use of the if (is_left).

You can do the following for example at the call place:

```
bounds.ForEachPointUntilImpl(is_left, other_point, non_intersecting_point_loop);
```

And inside the function you do:
```
  template <typename Func>
  void ForEachPointUntilImpl(bool is_left, LineString3d::const_iterator iter, Func&& f) {
    const LineString3d& points = is_left ? left_ : right_;
    for (auto iter = points.begin(); iter != points.end(); ++iter) {
      if (f(iter)) {
        break;
      }
    }
  }
```

Reply:
Good call. Done.